### PR TITLE
feat(DB-3584): add user managed resources

### PR DESCRIPTION
**Pull request by wmendes-ionos**
Adds user management resources to the provider.
It tests the controller unitarily given that user management must have a unique email and this workflow is a bit unrealistic via integration tests.

This PR adds:
- User Management

## Checklist

I have:

- [x] Add PR name as appropriate (e.g. `feat`/`fix`/`doc`/`test`/`refactor`)
- [x] Run `make reviewable` and `make crds.clean` to ensure the PR is ready for review
- [x] Add or update tests (if applicable)
- [x] Add or update Documentation using `make docs.update` (if applicable)
- [ ] Update `docs/CHANGELOG.md` file (label: `upcoming release`)
- [x] Check Sonar Cloud Scan
- [x] Update Github or Jira Issue

### CODE REVIEW

**Comment by cristiGuranIonos** [comment]
Can you add an example crd in examples/compute for the user and run make docs.update? Thanks!

**Comment by cristiGuranIonos** [comment]
We usually add integration tests. For user, we would normally add them to integration_tests_compute.sh.  Just mentioning this, they don't necessarily need to be added.

**Comment by wmendes-ionos** [comment]
@cristiGuranIonos make docs.update is updating a lot of other docs. is that intended? how to update only the user docs?

**Comment by wmendes-ionos** [comment]
@cristiGuranIonos regarding integration tests for users, this would require a random generator for the email, with that, emails cannot be reused anymore once the user is deleted, that is why I opted for unit tests.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:125]: possible nil pointer

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:158]: This code is duplicated, it also has nil exception

File: internal/clients/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/clients/compute/user/user.go:85]: This interface introduces strong coupling with UserPropertiesPost and UserPropertiesPut. Our team has no control over UserPropertiesPost or UserPropertiesPut, and you are forcing them to implement userPropsSetter.

File: internal/clients/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/clients/compute/user/user.go:97]: No need for this function to be public, it is used only in one file, and it can be private in that file.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:186]: do we need to have this wait for the Create method?

File: internal/controller/compute/user/user_test.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user_test.go:24]: No tests for Create, Update and Delete.

File: internal/controller/compute/user/user.go
Comment (wmendes-ionos) [reply internal/controller/compute/user/user.go:190]: Yes. The response code is 202. You can see that in our docs. [api.docs.ionos.com](https://api.ionos.com/docs/cloud/v6/)

File: internal/controller/compute/user/user_test.go
Comment (wmendes-ionos) [reply internal/controller/compute/user/user_test.go:24]: I said that in the PR description. WIP. tests are coming.

File: internal/controller/compute/user/user.go
Comment (wmendes-ionos) [reply internal/controller/compute/user/user.go]: It will help more if you can provide a concrete example or test case. Lines above check for `err` and the sdk will never return a `nil` User with a `nil` `err` at the same time. https://github.com/ionos-cloud/sdk-go/blob/master/api_user_management.go#L616-L746

File: internal/controller/compute/user/user.go
Comment (wmendes-ionos) [reply internal/controller/compute/user/user.go]: It will help more if you can provide a concrete example or test case. Lines above check for err and the sdk will never return a `nil` User with a `nil` `err` at the same time. https://github.com/ionos-cloud/sdk-go/blob/master/api_user_management.go#L616-L746

File: internal/clients/compute/user/user.go
Comment (wmendes-ionos) [reply internal/clients/compute/user/user.go:90]: Can you explain where is the strong coupling? Because the intention of the interface is exactly the opposite. We have the interface thus we control that. I have added a compile type check to make this even safer.

A bit more of context:
We create and update users. Both has the same set of properties however they are of different types on the sdk. Instead of having a duplicated function for each call, create(UserPropertiesPost) and update(UserPropertiesPut) this interface allow us to have one function to rule them all.

If you think this can be improved, please give a concrete example.

**Comment by cristiGuranIonos** [comment]
Hi, about docs just push the user md generated doc file.  The other files have the same values, but the order seems to be different. Will take a look as to why it happens.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [reply internal/controller/compute/user/user.go:198]: I see you added compute.WaitForRequest to the Create method, which was not there before. That what i was asking about.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [reply internal/controller/compute/user/user.go]: From the code it looks like props.GetS3CanonicalUserId() returns a pointer. if that is true, then this pointer can be nil, thus we need to check its value before using it.  
This is the method signature: func (o *UserProperties) GetS3CanonicalUserId() *string. This is similar to my other nil pointer comments.

File: internal/clients/compute/user/user.go
Comment (mshalabi-ionos) [reply internal/clients/compute/user/user.go:90]: when you introduce the interface you build an inheritance relationship. Inheritance is tightly coupled. The UserPropertiesPost and UserPropertiesPut are completely different objects, even they may have similar properties. The fix of this(similar properties) should be the SDK responsibility, and it could be done by composition not inheritance.

func setUserPropertiesPost(p v1alpha1.UserParameters, props UserPropertiesPost) { props.SetFirstname(p.FirstName) ...} and
func setUserPropertiesPut(p v1alpha1.UserParameters, props userPropsSetter) {props.SetFirstname(p.FirstName) ...} 
does not introduce duplication because they are dealing with completely different inputs.

My recommendation is to use two different methods instead of introducing coupling. 

I believe this can end with an arguable discussion, thus I will leave it to you to decide. If you still did not get my point, then we can talk about it.

Comment (wmendes-ionos) [reply internal/clients/compute/user/user.go:90]: I will choose to leave it as is.

File: internal/controller/compute/user/user.go
Comment (wmendes-ionos) [reply internal/controller/compute/user/user.go]: `S3CanonicalUserId` is computed by the ionos api. if the create request succeeds, the value will exists.

File: internal/clients/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/clients/compute/user/user.go:74]: possible nil exception in *props.GetAdministrator() also in other places

File: internal/clients/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/clients/compute/user/user.go:101]: This violates SOLID, interface should not know about its internal implementations.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [reply internal/controller/compute/user/user.go]: This is not what the contract says. It says it can be nil(func (o *UserProperties) GetS3CanonicalUserId() _*string_) if what you are saying is correct the contract will be func (o *UserProperties) GetS3CanonicalUserId() _string_

In all cases, this is a violation of defensive programming https://deviq.com/practices/defensive-programming

**Review by mshalabi-ionos — CHANGES_REQUESTED** [review]

File: internal/controller/compute/user/user.go
Comment (wmendes-ionos) [reply internal/controller/compute/user/user.go]: thanks fot the input, I have added a check.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:201]: possible nil in *observed.GetId() also in few lies down.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:265]: simple performance improvement
```suggestion
		if groupID != "" && !slices.Contains(forProviderGroups, groupID)   {
```

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:298]: if the items are in different orders it will think they are not equal. I did simple verification to verify it.
```suggestion
ipackage main

import (
	"fmt"
	"reflect"
)

func main() {
	first := []string{"100", "200", "300"}
	second := []string{"300", "100", "200"}
	third := []string{"100", "200", "300"}

	if !reflect.DeepEqual(first, second) {
		fmt.Println("not equal")
	} else {
		fmt.Println("equal")
	}

	if !reflect.DeepEqual(first, third) {
		fmt.Println("not equal")
	} else {
		fmt.Println("equal")
	}
}
```

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [inline internal/controller/compute/user/user.go:310]: possible nil pointer exception

File: apis/compute/v1alpha1/user_types.go
Comment (cristiGuranIonos) [inline apis/compute/v1alpha1/user_types.go:44]: Would be great if we could provide this from an env variable or secret. Something similar to what is being done for postgres_user, using CredentialsSource. Maybe in another CR.

Comment (cristiGuranIonos) [reply apis/compute/v1alpha1/user_types.go:44]: maybe we can add a todo comment to remember. thanks

**Review by mshalabi-ionos — COMMENTED** [review]
Please notice that I did not do any testing.

**Review by cristiGuranIonos — APPROVED** [review]

File: internal/clients/compute/user/user.go
Comment (wmendes-ionos) [reply internal/clients/compute/user/user.go:83]: Before we spoke on 06.02.2024 I had one function that addressed both methods.
https://github.com/ionos-cloud/crossplane-provider-ionoscloud/blob/d00d095099c861d2446e1bd1b468743885e19c72/internal/clients/compute/user/user.go

After talking to you about the importance of it and how it worked you still was against and said clearly to me that "It is ok to have duplication, they are of different classes"(using your own words. go not even has classes)

Now you telling me the same thing I told you? After changing the code again, you telling me this is duplication? You think I am an idiot or what?

File: apis/compute/v1alpha1/user_types.go
Comment (wmendes-ionos) [reply apis/compute/v1alpha1/user_types.go:44]: Super valid point. Thank you for raising this, I agree that is a much better way to address the password.

**Review by mshalabi-ionos — DISMISSED** [review]

File: internal/controller/compute/user/user.go
Comment (cristiGuranIonos) [inline internal/controller/compute/user/user.go:312]: While this is definitely better than the reflect.DeepEqual way, wouldn't slices.Equal that was added in go 1.21(i think?) solve this?

**Review by cristiGuranIonos — APPROVED** [review]

File: internal/utils/utils.go
Comment (amahdhaoui-ionos) [inline internal/utils/utils.go:31]: We can replace those by DereferenceOrZero[T]

```suggestion
func DereferenceOrZero[T any](v *T) T {
	if v == nil {
		return *new(T)
	}
	
	return *v
}
```

File: internal/controller/compute/user/user.go
Comment (amahdhaoui-ionos) [reply internal/controller/compute/user/user.go:312]: To run slices.Equal, we would need to run `slices.Sort` on both slices first. But I agree it would be a bit cleaner. This solution would be O(n log(n)) which is a bit more than Mohammed's O(n) comparison.
But we're working with fairly small slices were Mohammed's algo performs worst because of the `times 3` constant factor for allocating the sets and then iterating over them.

File: internal/controller/compute/user/user.go
Comment (amahdhaoui-ionos) [inline internal/controller/compute/user/user.go:283]: I'd rather use the `slices.Sort + slices.Equal` alternative, but if you want to use this one then it's good to note that it's unecessary to allocate boolean values to this map, a set can be allocated in go using a map[T]interface{}.

```suggestion
func createSet(s []string) map[string]interface{} {
	m := make(map[string]interface{})
	
	for i := range s {
		m[s[i]] = nil
	}
	
	return m
}
```

File: internal/controller/compute/user/user.go
Comment (amahdhaoui-ionos) [inline internal/controller/compute/user/user.go:138]: I saw you changed the if guards using these util functions. However, this doesn't replicate the same computational path.

Here you're mutating fields to their 0 value if the pointer is nil; whereas the previous implementation would not mutate the field if the pointer is nil. 

Not sure this behavior is what's intended.

**Review by amahdhaoui-ionos — APPROVED** [review]
Thanks for the clean PR @wmendes-ionos. 

NB @mshalabi-ionos: I've added a few comments. Please address them and ping me before merging.

**Review by amahdhaoui-ionos — APPROVED** [review]
Ran test, will proceed with merging the PR.

File: internal/utils/utils.go
Comment (cristiGuranIonos) [inline internal/utils/utils.go:17]: Thanks, this is definitely useful. We do have something similar defined [here](https://github.com/ionos-cloud/sdk-go/blob/7ff986147d3303a39c3cd3274931d1f02f5419d8/utils.go#L39) in the go sdk, but we need to update the version of the go sdk used in crossplane. Again, thanks for the contributions.

File: internal/controller/compute/user/user.go
Comment (mshalabi-ionos) [reply internal/controller/compute/user/user.go]: In the previous code, what the cr.Status.AtProvider.SecAuthActive will be in case of props.GetSecAuthActive()=nil?

**Pull request merged**
